### PR TITLE
Ensure we use the passed in libctx when loading EC private keys

### LINKED ISSUE #12150: d2i_AutoPrivateKey_ex does not use passed in OPENSSL_CTX for reading EC keys
Observed on master.

When using d2i_AutoPrivateKey_ex with a non-NULL library context, default lib context appears to still be used for some operations.

In particular, this has been noticed for reading EC private keys from non-encrypted PKCS8 format for some curves.

Reproduced by loading 'null' provider into NULL context, and trying to read EC keys using another context that includes 'default' provider. Should work, but doesn't.

```
#include <openssl/provider.h>
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/err.h>

int main(int argc, char *argv[]) 
{
    OPENSSL_CTX *otherctx = NULL;
    OSSL_PROVIDER *defctxnull = NULL;
    EVP_PKEY *pkey256, *pkey384, *pkey521 = NULL;
    size_t buflen256, buflen384, buflen521;
    int ret;
    const unsigned char *bufp256, *bufp384, *bufp521;
    bufp256 = OPENSSL_hexstr2buf("3041020100301306072A8648CE3D020106082A8648CE3D03010704273025020101042065070A12B4FE4C0C1BD9828C891D6BBA376E8D62B07034B8875786B9DBCC6F0F", &buflen256);
   
    bufp384 = OPENSSL_hexstr2buf("304E020100301006072A8648CE3D020106052B8104002204373035020101043073E33A05F2B6996D0C337F159E10A9174C0A825771137AAC46A25E1CE0C7B2F82040C227C8BE027E9669E004CB890B42", &buflen384);

    bufp521 = OPENSSL_hexstr2buf("3060020100301006072A8648CE3D020106052B81040023044930470201010442005C013EAF1833C11DA2FAD152F4D276682D29EC2C79229D05174989446365FB2CEDFBAF704C2685A8FCC09502690BC13F3A158AE6C54E0C0356F2E8AAE2D81A3351", &buflen521);

    otherctx = OPENSSL_CTX_new();
    if (otherctx == NULL)
       goto end;
    
    /* Prevent anything from using the default library context */
    defctxnull = OSSL_PROVIDER_load(NULL, "null");

    /* For info print out providers available */
    ret = OSSL_PROVIDER_available(otherctx, "default");
    printf("default provider available (otherctx) : %d\n", ret);
    ret = OSSL_PROVIDER_available(otherctx, "null");
    printf("null provider available (otherctx) : %d\n", ret);
    ret = OSSL_PROVIDER_available(NULL, "default");
    printf("default provider available (NULL) : %d\n", ret);
    ret = OSSL_PROVIDER_available(NULL, "null");
    printf("null provider available (NULL) : %d\n", ret);

    pkey256 = d2i_AutoPrivateKey_ex(NULL, &bufp256, buflen256, otherctx, NULL);
    if (pkey256 == NULL)
        printf("Failed to read P256 key\n");
    else
        printf("Succesfully read P256 EC private key\n");

    pkey384 = d2i_AutoPrivateKey_ex(NULL, &bufp384, buflen384, otherctx, NULL);
    if (pkey384 == NULL)
        printf("Failed to read P384 key\n");
    else
        printf("Succesfully read P384 EC private key\n");

    pkey521 = d2i_AutoPrivateKey_ex(NULL, &bufp521, buflen521, otherctx, NULL);
    if (pkey521 == NULL)
        printf("Failed to read P521 key\n");
    else
        printf("Succesfully read P521 EC private key\n");

end:
    ERR_print_errors_fp(stderr);
    EVP_PKEY_free(pkey256);
    EVP_PKEY_free(pkey384);
    EVP_PKEY_free(pkey521);
    OPENSSL_CTX_free(otherctx);
    OSSL_PROVIDER_unload(defctxnull);
   
    return 1;
}
```
Output when running:
```
default provider available (otherctx) : 1
null provider available (otherctx) : 0
default provider available (NULL) : 0
null provider available (NULL) : 1
Succesfully read P256 EC private key
Failed to read P384 key
Failed to read P521 key
801B1F6D4D7F0000:error::digital envelope routines:(unknown function):fetch failed:../crypto/evp/evp_fetch.c:307:Default library context, Algorithm (AES-256-ECB), Properties ()
801B1F6D4D7F0000:error::digital envelope routines:(unknown function):fetch failed:../crypto/evp/evp_fetch.c:307:Default library context, Algorithm (AES-256-CTR), Properties ()
801B1F6D4D7F0000:error::random number generator:(unknown function):error initialising drbg:../crypto/rand/drbg_lib.c:404:
801B1F6D4D7F0000:error::elliptic curve routines:(unknown function):ladder pre failure:../crypto/ec/ec_mult.c:278:
801B1F6D4D7F0000:error::elliptic curve routines:(unknown function):decode error:../crypto/ec/ec_ameth.c:212:
801B1F6D4D7F0000:error::digital envelope routines:(unknown function):private key decode error:../crypto/evp/evp_pkey.c:50:
801B1F6D4D7F0000:error::elliptic curve routines:(unknown function):ladder pre failure:../crypto/ec/ec_mult.c:278:
801B1F6D4D7F0000:error::elliptic curve routines:(unknown function):decode error:../crypto/ec/ec_ameth.c:212:
801B1F6D4D7F0000:error::digital envelope routines:(unknown function):private key decode error:../crypto/evp/evp_pkey.c:50:
```

### PR DESCRIPTION
**Pull request by mattcaswell**
If the public key is not present in an PKCS8 EC private key we can end up not using the libctx.

Fixes #12150

### CODE REVIEW

File: crypto/ec/ec_ameth.c [hunk 1]
```diff
@@ -202,22 +210,22 @@ static int eckey_priv_decode(EVP_PKEY *pkey, const PKCS8_PRIV_KEY_INFO *p8)
...
-    eckey = eckey_type2param(ptype, pval);
+    eckey = eckey_type2param(ptype, pval, libctx, propq);
 
     if (eckey == NULL)
         goto ecliberr;
 
     /* We have parameters now set private key */
     if (!d2i_ECPrivateKey(&eckey, &p, pklen)) {
-        ECerr(EC_F_ECKEY_PRIV_DECODE, EC_R_DECODE_ERROR);
+        ECerr(EC_F_ECKEY_PRIV_DECODE_WITH_LIBCTX, EC_R_DECODE_ERROR);
```
Comment (slontis) [inline crypto/ec/ec_ameth.c:220]: why bother propagating these flags - just use zero.

File: crypto/ec/ec_curve.c [hunk 2]
```diff
@@ -3288,14 +3289,15 @@ static EC_GROUP *ec_group_new_from_data(OPENSSL_CTX *libctx,
...
+                                                 const char *propq, int nid)
 {
     EC_GROUP *ret = NULL;
     const ec_list_element *curve;
 
     if ((curve = ec_curve_nid2curve(nid)) == NULL
-        || (ret = ec_group_new_from_data(libctx, *curve)) == NULL) {
-        ECerr(EC_F_EC_GROUP_NEW_BY_CURVE_NAME_EX, EC_R_UNKNOWN_GROUP);
+        || (ret = ec_group_new_from_data(libctx, propq, *curve)) == NULL) {
+        ECerr(EC_F_EC_GROUP_NEW_BY_CURVE_NAME_WITH_LIBCTX, EC_R_UNKNOWN_GROUP);
```
Comment (slontis) [inline crypto/ec/ec_curve.c:3300]: same here with the flags

File: crypto/ec/ec_key.c [hunk 3]
```diff
@@ -631,6 +639,11 @@ OPENSSL_CTX *ec_key_get_libctx(const EC_KEY *key)
     return key->libctx;
 }
 
+const char *ec_key_get_propq(const EC_KEY *key)
```
Comment (slontis) [inline crypto/ec/ec_key.c:642]: get0

File: crypto/ec/ec_lib.c [hunk 4]
```diff
@@ -23,7 +23,8 @@
 
 /* functions for EC_GROUP objects */
 
-EC_GROUP *ec_group_new_ex(OPENSSL_CTX *libctx, const EC_METHOD *meth)
```
Comment (slontis) [inline crypto/ec/ec_lib.c:26]: with_libctx?

File: crypto/ec/ec_local.h [hunk 5]
```diff
@@ -593,10 +595,12 @@ int ec_group_simple_order_bits(const EC_GROUP *group);
  *  Creates a new EC_GROUP object
  *  \param   libctx The associated library context or NULL for the default
  *                  library context
+ *  \param   propq  Any property query string
  *  \param   meth   EC_METHOD to use
  *  \return  newly created EC_GROUP object or NULL in case of an error.
  */
-EC_GROUP *ec_group_new_ex(OPENSSL_CTX *libctx, const EC_METHOD *meth);
+EC_GROUP *ec_group_new_ex(OPENSSL_CTX *libctx, const char *propq,
```
Comment (slontis) [inline crypto/ec/ec_local.h:602]: with_libctx

**Comment by slontis** [comment]
Travis errors are valid.

**Comment by slontis** [comment]
Code looks good - This libctx really spreads quickly :)

**Comment by mattcaswell** [comment]
All feedback addressed, and the travis error (hopefully) fixed.

Please take another look.

**Comment by t8m** [comment]
@mattcaswell Did you really mean to close it?

**Comment by mattcaswell** [comment]
> @mattcaswell Did you really mean to close it?

Oops! No - thanks. Reopened.

**Comment by slontis** [comment]
make update fails.. (more error flags should be zero with the name change).

File: crypto/ec/ec_lib.c [hunk 6]
```diff
@@ -23,8 +23,8 @@
 
 /* functions for EC_GROUP objects */
 
-EC_GROUP *ec_group_new_ex(OPENSSL_CTX *libctx, const char *propq,
-                          const EC_METHOD *meth)
+EC_GROUP *ec_group_new_with_libctx(OPENSSL_CTX *libctx, const char *propq,
```
Comment (slontis) [inline crypto/ec/ec_lib.c:26]: Fixup the ECerr() in this function

**Comment by mattcaswell** [comment]
Rebased, and "make update" issues fixed.

File: util/libcrypto.num [hunk 7]
```diff
@@ -4683,9 +4683,9 @@ ERR_set_error                           ?	3_0_0	EXIST::FUNCTION:
 ERR_vset_error                          ?	3_0_0	EXIST::FUNCTION:
 X509_get0_authority_issuer              ?	3_0_0	EXIST::FUNCTION:
 X509_get0_authority_serial              ?	3_0_0	EXIST::FUNCTION:
-EC_GROUP_new_by_curve_name_ex           ?	3_0_0	EXIST::FUNCTION:EC
-EC_KEY_new_ex                           ?	3_0_0	EXIST::FUNCTION:EC
-EC_KEY_new_by_curve_name_ex             ?	3_0_0	EXIST::FUNCTION:EC
+EC_GROUP_new_by_curve_name_ex           ?	3_0_0	NOEXIST::FUNCTION:EC
+EC_KEY_new_ex                           ?	3_0_0	NOEXIST::FUNCTION:EC
+EC_KEY_new_by_curve_name_ex             ?	3_0_0	NOEXIST::FUNCTION:EC
```
Comment (t8m) [inline util/libcrypto.num:4688]: Should these lines be removed completely?

Comment (slontis) [reply util/libcrypto.num:4688]: Dont they get removed when a release happens?

Comment (mattcaswell) [reply util/libcrypto.num:4688]: I expect all of these to get cleaned up when we do the final release

Comment (t8m) [reply util/libcrypto.num:4688]: ok

**Comment by mattcaswell** [comment]
Pushed. Thanks! I resolved a trivial conflict in libcrypto.num.

**Pull request closed without merging**